tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd

```diff
@@ -0,0 +1,584 @@
+
+
+## Number of components
+
+A direct question arises from this is: How many dimensions do we want to keep in order to represent the latent structure in the data?
+This is a very central question for all latent factor models, i.e. how many dimensions are needed to represent that data in the latent space.
+In the case of PCA, there exist a lot of heuristics to make that choice.
+By using the pPCA model, this can be accomplished very elegantly, with a technique called *Automatic Relevance Determination*(ARD).
+Essentially, we are using a specific prior over the factor loadings W that allows us to prune away dimensions in the
+latent space. The prior is determined by a precision hyperparameter $\alpha$. Here, smaller values of $\alpha$ correspond to more important components.
+You can find more details about this in the Bishop book mentioned in the introduction.
+
+
+```julia
+@model function pPCA_ARD(x, ::Type{TV} = Array{Float64}) where {TV}
+  # Dimensionality of the problem.
+  N, D = size(x)
+
+  # latent variable z
+  z ~ filldist(Normal(), D, N)
+
+  # weights/loadings w with Automatic Relevance Determination part
+  alpha ~ filldist(Gamma(1.0, 1.0), D)
+  w ~ filldist(MvNormal(zeros(D), 1. ./ sqrt.(alpha)), D)
+
+  mu = (w' * z)'
+
+  tau ~ Gamma(1.0, 1.0);
+  for d in 1:D
+    x[:,d] ~ MvNormal(mu[:, d], 1. / sqrt(tau))
+  end
+end;
+```
+```julia
+ppca_ARD = pPCA_ARD(expression_matrix)
+chain_pccaARD = sample(ppca_ARD, NUTS(), 500)
+
+StatsPlots.plot(group(chain_pccaARD, :alpha))
+```
+
+Here we look at the convergence of the chains for the $\alpha$ parameter. This parameter determines the relevance of individual components. We can see that the chains have converged and the posterior of the alpha parameters is centered around much smaller values in two instances. Below, we will use the mean of the small values to select the *relevant* dimensions - we can clearly see based on the values of $\alpha$ that there should be two dimensions in this example.
+
+```julia
+# Extract paramter estimates for plotting - mean of posterior
+w = permutedims(reshape(mean(group(chain_pccaARD, :w))[:,2], (n_genes,n_genes)))
+z = permutedims(reshape(mean(group(chain_pccaARD, :z))[:,2], (n_genes, n_cells)))'
+α = mean(group(chain_pccaARD, :alpha))[:,2]
+α
+
+```
+We can inspect alpha to see which elements are small, i.e. have a high relevance.
+
+```julia
+alpha_indices = sortperm(α)[1:2]
+X = w[alpha_indices, alpha_indices] * z[alpha_indices,:]
+
+df_rec = DataFrame(X', :auto)
+df_rec[!,:cell] = 1:n_cells
+DataFrames.stack(df_rec, 1:2) |> @vlplot(:rect, "cell:o", "variable:o", color=:value)
+
+df_pre = DataFrame(z', :auto)
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:n_genes)]))
+df_pre[!,:cell] = 1:n_cells
+
+DataFrames.stack(df_pre, 1:n_genes) |> @vlplot(:rect, "cell:o", "variable:o", color=:value)
+
+df_pre[!,:type] = repeat([1, 2], inner = n_cells÷2)
```

Comment:
```suggestion
df_pre[!,:type] = repeat([1, 2], inner=n_cells÷2)
```
You're using no whitespace around equal symbols in keyword arguments pretty consistently.